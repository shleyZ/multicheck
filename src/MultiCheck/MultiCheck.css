/* TODO */
.MultiCheck {
}

.MultiCheckLabel{
  background-color: #DADADA;
  padding: 8px;
  font-weight: bold;
}

.MultiCheckContentWrap{
  background-color: #F7F7F9;
  padding: 8px;
}

.MultiCheckContent{
  border: 1px solid #E5E5E5;
}

.ColumnItem{
  display: inline-block;
}