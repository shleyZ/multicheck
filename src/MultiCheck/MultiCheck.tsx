import React, { useEffect, useState } from 'react';
import CheckBox from './components/CheckBox';
import './MultiCheck.css';

export type Option = {
  label: string,
  value: string
}

/**
 * Notice:
 * 1. There should be a special `Select All` option with checkbox to control all passing options
 * 2. If columns > 1, the options should be placed from top to bottom in each column
 *
 * @param {string} label - the label text of this component
 * @param {Option[]} options - options
 * @param {string[]} values - default checked option values
 * @param {number} columns - default value is 1
 * @param {Function} onChange - when checked options are changed,
 *                             they should be passed to outside
 */
type Props = {
  label?: string,
  options: Option[],
  columns?: number,
  values?: string[],
  onChange?: (options: Option[]) => void,
}

const MultiCheck: React.FunctionComponent<Props> = (props): JSX.Element => {
  const { label, options, values, onChange, columns } = props;
  const [checked, setChecked] = useState<string[]>([]);
  const [selectAllChecked, setSelectAllChecked] = useState<boolean>(false);

  const onAllChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
    if(selectAllChecked) {
      setChecked([])
    } else {
      const allValues = [] as string[];
      options.forEach((item: Option) => allValues.push(item.value))
      setChecked(allValues)
    }
    setSelectAllChecked(!selectAllChecked)
  }

  const onItemChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
    const strValue = String(e.target.value);
    const currentChecked = [...checked];
    const index = currentChecked?.indexOf(strValue)
    if(index === -1) {
      currentChecked.push(strValue)
    } else {
      currentChecked?.splice(index, 1)
    }
    setChecked(currentChecked)
    setSelectAllChecked(currentChecked.length === options.length)
  }

  // initialized checked state
  useEffect(() => {
    setChecked(values || []);
    if(values && values.length === options.length) {
      setSelectAllChecked(true)
    } else {
      setSelectAllChecked(false)
    }
  }, []);

  // Monitor the checked state and react to the higher-level component
  useEffect(() => {
    if(onChange) {
      const checkedOptions = [] as Option[];
      options.forEach((option) => {
        if(checked.indexOf(option.value) !== -1) {
          checkedOptions.push(option)
        }
      })
      onChange(checkedOptions);
    }
  }, [checked])

  // counts per columns
  const columnCounts = Math.ceil(options.length / (columns || 1));
  // column percent %
  const columnWidths = 1/(columns || 1) * 100;

  const selesctAll = (
    <CheckBox 
      checked={!!selectAllChecked} 
      label="Select All"
      value="all"
      key="all"
      handleChanged={onAllChanged}
      />
    )
  /**
   * render column by index
   * @param {number} index - column index
   */
  const renderColumn = (index: number) => {
    const columnDom = [];
    if(index === 0) {
      columnDom.push(selesctAll)
      options.slice(0, columnCounts-1).map((opt: Option) => {
        columnDom.push(
          <CheckBox
            checked={checked.indexOf(String(opt.value)) !== -1}
            key={opt.value}
            label={opt.label}
            value={opt.value}
            handleChanged={onItemChanged}
          />
        )
      })
    } else {
      const startIndex = (index - 1)*columnCounts + (columnCounts - 1)
      options.slice(startIndex, columnCounts+startIndex).map((opt: Option) => {
        columnDom.push(
          <CheckBox
            checked={checked.indexOf(String(opt.value)) !== -1}
            key={opt.value}
            label={opt.label}
            value={opt.value}
            handleChanged={onItemChanged}
          />
        )
      })
    }
    return (
      <div 
        key={index}
        style={{
          width: `${columnWidths}%`, 
          display: 'inline-block'
        }}
        >
        {columnDom}
      </div>
    );
  }

  /**
   * render all columns
   * @param {number} columns 
   */
  const renderColumns = (columns: number) => {
    const res = [];
    for(let i = 0; i < columns; i++) {
      const columnDom = renderColumn(i)
      res.push(columnDom)
    }
    return res;
  }
  
  return (
    <div className='MultiCheck'>
      {/* TODO */}
      <div className="MultiCheckLabel">
        <label>{label}</label>
      </div>
      <div className="MultiCheckContentWrap">
        <div className="MultiCheckContent">
          {renderColumns(columns || 1)}
        </div>
      </div>
    </div>
  )
}

export default MultiCheck;
