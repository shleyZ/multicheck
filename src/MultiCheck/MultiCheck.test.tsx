
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jsdom-global/register';
import React from 'react';
import MultiCheck from './MultiCheck';
import type { Option } from './MultiCheck';
// TODO more tests

configure({adapter: new Adapter()});

describe('MultiCheck', () => {
  describe('initialize', () => {
    it('renders the label if label provided', (done) => {
      // TODO
      const wrapper = mount(
        <MultiCheck label='my-multi-check' options={[]} />
      );
      expect(wrapper.html()).toContain('<label>my-multi-check</label>');
      done()
    });

    it('renders the columns if columns is not provided', (done) => {
      const options: Option[] = [
        {label: 'aaa', value: '111',},
        {label: 'bbb', value: '222',},
        {label: 'ccc', value: '333',},
        {label: 'ddd', value: '444',},
        {label: 'eee', value: '555',},
        {label: 'fff', value: '666',},
        {label: 'ggg', value: '777',},
        {label: 'hhh', value: '888',},
        {label: 'iii', value: '999',},
      ]
      const wrapper = mount(
        <MultiCheck label='my-multi-check' options={options}/>
      );
      expect(wrapper.find('.MultiCheckContent').children().length).toBe(1);
      done()
    });

    it('renders the columns if columns provided', (done) => {
      const options: Option[] = [
        {label: 'aaa', value: '111',},
        {label: 'bbb', value: '222',},
        {label: 'ccc', value: '333',},
        {label: 'ddd', value: '444',},
        {label: 'eee', value: '555',},
        {label: 'fff', value: '666',},
        {label: 'ggg', value: '777',},
        {label: 'hhh', value: '888',},
        {label: 'iii', value: '999',},
      ]
      const wrapper = mount(
        <MultiCheck label='my-multi-check' options={options} columns={6}/>
      );
      expect(wrapper.find('.MultiCheckContent').children().length).toBe(6);
      done()
    });
  });

  describe('snapshot', () => {
    it('get snapshot', () => {
      const wrapper = mount(
        <MultiCheck label='my-multi-check' options={[]} />
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    })
  })
});
