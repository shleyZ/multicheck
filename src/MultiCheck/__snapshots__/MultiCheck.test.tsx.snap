// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCheck snapshot get snapshot 1`] = `
<MultiCheck
  label="my-multi-check"
  options={Array []}
>
  <div
    className="MultiCheck"
  >
    <div
      className="MultiCheckLabel"
    >
      <label>
        my-multi-check
      </label>
    </div>
    <div
      className="MultiCheckContentWrap"
    >
      <div
        className="MultiCheckContent"
      >
        <div
          key="0"
          style={
            Object {
              "display": "inline-block",
              "width": "100%",
            }
          }
        >
          <CheckBoxItem
            checked={false}
            handleChanged={[Function]}
            key="all"
            label="Select All"
            value="all"
          >
            <label
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <input
                checked={false}
                onChange={[Function]}
                type="checkbox"
                value="all"
              />
              <span>
                Select All
              </span>
            </label>
          </CheckBoxItem>
        </div>
      </div>
    </div>
  </div>
</MultiCheck>
`;
